File: ADT/stack-queue-pblm/20.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
    const st = [];

    for (const ch of s) {
        if (st.length === 0 || ch === '(' || ch === '{' || ch === '[') {
            st.push(ch);
            continue;
        }

        if (ch === ')') {
            if (st[st.length - 1] === '(')
                st.pop();
            else
                return false;
        } else if (ch === '}') {
            if (st[st.length - 1] === '{')
                st.pop();
            else
                return false;
        } else if (ch === ']') {
            if (st[st.length - 1] === '[')
                st.pop();
            else
                return false;
        }
    }

    return st.length === 0;
};



/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
    const st = []
    for (const ch of s) {
        if (ch === '(') {
            st.push(')')
        } else if (ch === '{') {
            st.push('}')
        } else if (ch === '[') {
            st.push(']')
        } else if (st.length === 0 || st[st.length - 1] !== ch) {
            return false
        } else {
            st.pop()
        }
    }
    return st.length === 0
};

File: ADT/stack-queue-pblm/20.py
class Solution:
    def isValid(self, s: str) -> bool:
        st = []

        for ch in s:
            if not st or ch in '({[':
                st.append(ch)
                continue

            if ch == ')':
                if st[-1] == '(':
                    st.pop()
                else:
                    return False
            elif ch == '}':
                if st[-1] == '{':
                    st.pop()
                else:
                    return False
            elif ch == ']':
                if st[-1] == '[':
                    st.pop()
                else:
                    return False

        return not st


class Solution:
    def isValid(self, s: str) -> bool:
        st = []
        
        for ch in s:
            if ch == '(':
                st.append(')')
            elif ch == '{':
                st.append('}')
            elif ch == '[':
                st.append(']')
            elif not st or st[-1] != ch:
                return False
            else:
                st.pop()
        
        return not st